### PR TITLE
Add comms-vm

### PR DESCRIPTION
**Pull request by shamma-alblooshi1**
## Description of changes
This PR modifies the existing element-vm into a more generic comms-vm and adds Slack to it. Slack runs as a PWA due to 
issues in the electron package. [(SP-4808)](https://ssrc.atlassian.net/browse/SP-4808) 

## Checklist for things done

- [X] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [X] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [x] PR linked to architecture documentation and requirement(s) (ticket id)
- [ ] Test procedure described (or includes tests). Select one or more:
  - [ ] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [x] Author has run `nix flake check --accept-flake-config` and it passes
- [x] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

## Testing
Open the element application. Also open the slack application. Both applications should be able to launch. Sign in into slack, to make sure everything is well. ( Make sure you are connected to wifi first to test the signing in :) ) 
Known issues:
You need to wait for some time till the wifi connects to be able to launch slack ( not a slack problem, but i believe ghaf needs some time to just connect)

### CODE REVIEW

File: modules/reference/appvms/comms.nix [hunk 1]
```diff
@@ -16,13 +16,33 @@
...
       pkgs.element-desktop
       pkgs.element-gps
+      pkgs.slack
       pkgs.gpsd
       pkgs.tcpdump
       pkgs.pulseaudio
+      pkgs.chromium
+      pkgs.xdg-utils
+      pkgs.eolie #web bowser
+      pkgs.xorg.xinit # Add xinit to start X server
```
[the remark this answers is not in the document]
Comment (shamma-alblooshi1) [reply modules/reference/appvms/comms.nix]: Oh yeah i was supposed to remove it

**Comment by leivos-unikie** [comment]
Hostname change to comm-vm has not succeeded well:
- `microvm -l` shows still element-vm
- `ssh comm-vm` works but then we end up to `[ghaf@element-vm:~]$`
- and `hostname` in element-vm shows 'element-vm'

**Comment by leivos-unikie** [comment]
Sing in/out from Slack works.

Got one 'WebKit web engine crashed' from Workspace -> Toos & settings -> Workspace analytics

File: modules/common/services/desktop.nix [hunk 2]
```diff
@@ -93,10 +93,22 @@ in {
...
             }
 
+            {
+              name = "Slack";
+              path = "${pkgs.openssh}/bin/ssh -i ${privateSshKeyPath} -o StrictHostKeyChecking=no comms-vm run-waypipe slack --enable-features=UseOzonePlatform --ozone-platform=wayland";
+              icon = "${pkgs.icon-pack}/slack.svg";
+            }
+
+            {
+              name = "Browser";
```
Comment (brianmcgillion) [inline modules/common/services/desktop.nix:107]: do not expose the browser here, as this is not the browser VM

File: modules/reference/appvms/comms.nix [hunk 3]
```diff
@@ -14,15 +14,34 @@
...
+    };
+  in
     [
       pkgs.element-desktop
       pkgs.element-gps
+      pkgs.slack
       pkgs.gpsd
       pkgs.tcpdump
       pkgs.pulseaudio
+      pkgs.chromium
```
Comment (brianmcgillion) [inline modules/reference/appvms/comms.nix:40]: why chromium?

**Review by brianmcgillion — CHANGES_REQUESTED** [review]

**Comment by shamma-alblooshi1** [comment]
> Hostname change to comms-vm has not succeeded well:
> 
> * `microvm -l` shows still element-vm
> * `ssh comms-vm` works but then we end up to `[ghaf@element-vm:~]$`
> * and `hostname` in element-vm shows 'element-vm'

This is fixed now!

File: modules/reference/appvms/comms.nix (hunk 3 above)
Comment (shamma-alblooshi1) [reply modules/reference/appvms/comms.nix:40]: i was testing the chromium first, i'll remove it

File: packages/icon-pack/default.nix [hunk 4]
```diff
@@ -17,7 +17,9 @@
     "firefox.svg"
     "microsoft-365.svg"
     "ms-outlook.svg"
+    "org.gnome.Eolie.svg"
```
Comment (brianmcgillion) [inline packages/icon-pack/default.nix:20]: remove this as you are not exporting it to the desktop UI

**Review by brianmcgillion — APPROVED** [review]

**Comment by shamma-alblooshi1** [comment]
> Sing in/out from Slack works.
> 
> Got one 'WebKit web engine crashed' from Workspace -> Toos & settings -> Workspace analytics

Replaced the webkit web engine with chromium to not get this .

**Comment by leivos-unikie** [comment]
Tested on Lenovo-X1
- All apps launch
- Sign in works in Element and Slack
- ci-test-automation pass (after modifying element-vm -> comms-vm), performance ok

If Slack window is closed before successfully logging in it does not launch anymore.

If signing in with email in "Check you email for a code" phase Slack offers links: "Open Gmail" and "Open Outlook". Clicking those links leads to the "untrusted" browser. Would it be difficult to set at least the "Open Outlook" link to open the Outlook App or Outlook sign in page in the "trusted" browser?

**Comment by shamma-alblooshi1** [comment]
> Tested on Lenovo-X1
> 
> * All apps launch
> * Sign in works in Element and Slack
> * ci-test-automation pass (after modifying element-vm -> comms-vm), performance ok
> 
> If Slack window is closed before successfully logging in it does not launch anymore.
> 
> If signing in with email in "Check you email for a code" phase Slack offers links: "Open Gmail" and "Open Outlook". Clicking those links leads to the "untrusted" browser. Would it be difficult to set at least the "Open Outlook" link to open the Outlook App or Outlook sign in page in the "trusted" browser?

I thought of this from the very beginning, where the sign in would be in the untrusted or trusted browser , but its not very straightforward because you would need to re-direct back to slack app ( its not as as simple as the pdf handler where there is no redirection after), so we decided for now handle the signing in in the comms vm for now ,and later will focus on doing the proper offloading to the remote browser next.

**Comment by leivos-unikie** [comment]
> If Slack window is closed before successfully logging in it does not launch anymore.

The consequent launch attempts stop to this in comms-vm journalctl:
```
Aug 15 14:34:25 comms-vm sshd-session[1223]: Accepted publickey for ghaf from 192.168.100.3 port 39352 ssh2: ED25519 SHA256:xDZvDND9Aap37sS+aziDmD4szQ5ibGs8bZPpydtDakg
Aug 15 14:34:25 comms-vm sshd-session[1223]: pam_unix(sshd:session): session opened for user ghaf(uid=1000) by (uid=0)
Aug 15 14:34:25 comms-vm sshd-session[1223]: pam_systemd(sshd:session): New sd-bus connection (system-bus-pam-systemd-1223) opened.
Aug 15 14:34:25 comms-vm systemd-logind[650]: New session 4 of user ghaf.
Aug 15 14:34:25 comms-vm systemd[1]: Started Session 4 of User ghaf.
```

whereas at the first launch it goes further:
```
Aug 15 14:33:48 comms-vm sshd-session[863]: Accepted publickey for ghaf from 192.168.100.3 port 33236 ssh2: ED25519 SHA256:xDZvDND9Aap37sS+aziDmD4szQ5ibGs8bZPpydtDakg
Aug 15 14:33:48 comms-vm sshd-session[863]: pam_unix(sshd:session): session opened for user ghaf(uid=1000) by (uid=0)
Aug 15 14:33:48 comms-vm sshd-session[863]: pam_systemd(sshd:session): New sd-bus connection (system-bus-pam-systemd-863) opened.
Aug 15 14:33:48 comms-vm systemd-logind[650]: New session 3 of user ghaf.
Aug 15 14:33:48 comms-vm systemd[1]: Started Session 3 of User ghaf.
Aug 15 14:33:51 comms-vm systemd[750]: Created slice User Core Session Slice.
Aug 15 14:33:51 comms-vm systemd[750]: Starting D-Bus User Message Bus...
Aug 15 14:33:51 comms-vm systemd[750]: Started D-Bus User Message Bus.
Aug 15 14:33:51 comms-vm dbus-daemon[627]: [system] Activating via systemd: service name='org.freedesktop.PolicyKit1' unit='polkit.service' requested by ':1.2' (uid=0 pid=650 comm="/nix/store/ki9sp0hrj28f9fzkvwrvy06yqh50yd3l-appvm-" label="kernel")
Aug 15 14:33:51 comms-vm systemd[1]: Starting Authorization Manager...
Aug 15 14:33:51 comms-vm polkitd[950]: Started polkitd version 124
Aug 15 14:33:51 comms-vm polkitd[950]: Loading rules from directory /etc/polkit-1/rules.d
Aug 15 14:33:51 comms-vm polkitd[950]: Loading rules from directory /run/current-system/sw/share/polkit-1/rules.d
Aug 15 14:33:52 comms-vm polkitd[950]: Finished loading, compiling and executing 3 rules
Aug 15 14:33:52 comms-vm dbus-daemon[627]: [system] Successfully activated service 'org.freedesktop.PolicyKit1'
Aug 15 14:33:52 comms-vm polkitd[950]: Acquired the name org.freedesktop.PolicyKit1 on the system bus
Aug 15 14:33:52 comms-vm systemd[1]: Started Authorization Manager.
Aug 15 14:34:05 comms-vm /nix/store/cnkn[1073]: Opening https://app.slack.com/ssb/signin?ssb_vid=f8c642a6aaa1dd825d4c52685b3d875b&ssb_instance_id=b43d2a9d-b786-4a4b-ba68-4276ec672e33&v=4.39.90&from_desktop_app=1
```

**Comment by leivos-unikie** [comment]
I monitored the process side in comms-vm while launching and closing Slack
`watch "ps aux | grep slack"`

Connected Wifi, launched slack and signed in. 
- If the slack window is closed in signed-in state either by clicking [x] or File -> Close only the UI window closes but slack related processes in comms-vm don't stop. In this case Slack can be reopened in its signed-in state by clicking the icon.
- On the other hand File -> Quit Slack shuts down also the processes in comms-vm. In this case Slack don't anymore start by clicking its icon. Need to do `sudo systemctl restart microvm@comms-vm.service` on ghaf-host or reboot first.

If Slack is not in signed-in state pressing [x] or File -> Close shuts down both UI and the slack processes. Clicking Slack icon restarts the processes but Slack window doesn't show up anymore.

If user happens to try to launch Slack without WiFi connected it takes quite long (~25 sec) for the Slack window (with Slack cannot connect) to show up.

**Comment by shamma-alblooshi1** [comment]
> * In this case Slack can be reopened in its si



> I monitored the process side in comms-vm while launching and closing Slack `watch "ps aux | grep slack"`
> 
> Connected Wifi, launched slack and signed in.
> 
> * If the slack window is closed in signed-in state either by clicking [x] or File -> Close only the UI window closes but slack related processes in comms-vm don't stop. In this case Slack can be reopened in its signed-in state by clicking the icon.
> * On the other hand File -> Quit Slack shuts down also the processes in comms-vm. In this case Slack doesn't start anymore by clicking its icon. Need to do `sudo systemctl restart microvm@comms-vm.service` on ghaf-host or reboot first.
> 
> If Slack is not in signed-in state pressing [x] or File -> Close shuts down both UI and the slack processes. Clicking Slack icon restarts the processes but Slack window doesn't show up anymore.
> 
> If user happens to try to launch Slack without WiFi connected it takes quite long (~25 sec) for the Slack window (with "Slack cannot connect") to show up.

Not sure what the problem is , will try to sort it out

**Comment by brianmcgillion** [comment]
the issue here is that the X or close of the window are just closig the wayland display of the application and is not actually stopping the application. Just removing the window for that application. It would require that the application is actually shutdown when the wayland display is closed. OR that when you click the icon again there is a "resume" UI operation to reconnect to the running Slack.

Slack is also bad at shutting down. When you close the slack window on a normal Linux desktop (NixOS atleast), the slack app is not closed. I generally have to do some form of `killall` or `kill -9 ...` to really close slack

**Comment by leivos-unikie** [comment]
Slack app can now be opened and closed multiple times without problems.

Also in case there is no WiFi connection Slack window opens within few seconds. However it stays blank very long, for about 90 sec until saying "This site can't be reached...". Could that timeout in trying connection be shortened?

**Review by leivos-unikie — CHANGES_REQUESTED** [review]
Shutdown/Reboot buttons have stopped working!

**Review by leivos-unikie — APPROVED** [review]

File: modules/reference/services/dendrite-pinecone/dendrite-config.nix [hunk 5]
```diff
@@ -22,9 +22,7 @@
         in
         "${lib.head vmNetworking.networking.nat.internalInterfaces}";
 
-      getElementVmEntry = builtins.filter (
-        x: x.name == "element-vm"
-      ) config.ghaf.networking.hosts.entries;
+      getElementVmEntry = builtins.filter (x: x.name == "comms-vm") config.ghaf.networking.hosts.entries;
```
[the remark this answers is not in the document]
Comment (shamma-alblooshi1) [reply modules/reference/services/dendrite-pinecone/dendrite-config.nix:25]: to what ?

**Review by brianmcgillion — APPROVED** [review]

**Review by brianmcgillion — APPROVED** [review]

**Comment by leivos-unikie** [comment]
Re-tested on Lenovo-X1

- Element launches and sing in works.
- Slack launch and sing in works. Closing and launching multiple times does not cause problems.
- ci-test-automation passed

**Comment by brianmcgillion** [comment]
rebased for the givc changes

**Pull request merged**